app/api/ada/adaTransactions/adaTransactionsHistory.js

```diff
@@ -40,116 +46,165 @@ declare var CONFIG : ConfigType;
       _updateAdaTxsHistoryForGroupOfAddresses([], groupOfAddresses, dateFrom, addresses)
     ));
+
+    // save transactions in lovefieldDB
     return saveTxs(mappedTxs);
   } catch (error) {
     Logger.error('adaTransactionsHistory::updateAdaTxsHistory error: ' + stringifyError(error));
     throw new UpdateAdaTxsHistoryError();
   }
 }
 
-async function _getTxsForChunksOfAddresses(addresses, apiCall) {
+/** Split API call on list of addresses into batched requests */
+async function _getTxsForChunksOfAddresses(
+  addresses: Array<string>,
+  apiCall: (Array<string>) => Promise<Array<AdaTransaction>>
+): Promise<Array<AdaTransaction>> {
   const groupsOfAddresses = _.chunk(addresses, addressesLimit);
   const groupedTxsPromises = groupsOfAddresses.map(apiCall);
   const groupedTxs = await Promise.all(groupedTxsPromises);
   return groupedTxs.reduce((accum, chunkTxs) => accum.concat(chunkTxs), []);
 }
 
+/** Recursively fetch transaction history for a set of addresses from backend API
+  * At the same time, update best block saved in local storage
+*/
 async function _updateAdaTxsHistoryForGroupOfAddresses(
-  previousTxs,
-  groupOfAddresses,
-  dateFrom,
-  allAddresses
-) {
-  const mostRecentTx = getMostRecentTx(previousTxs);
-  const updatedDateFrom = mostRecentTx ? moment(mostRecentTx.ctMeta.ctmUpdate) : dateFrom;
+  previousTxs: Array<AdaTransaction>,
+  groupOfAddresses: Array<string>,
+  dateFrom: Date,
+  allAddresses: Array<string>
+): Promise<Array<AdaTransaction>> {
+  /* We want to get the transaction history of a list of addresses
+   * Howevever, the backend API has a limited size in its response
+   * Therefore we keep calling it repeatedly until we get the full history.
+  */
+
+  // Move cutoff date forward to make progress on recursive calls
+  const updatedDateFrom = previousTxs.length > 0
+    ? previousTxs[previousTxs.length - 1].ctMeta.ctmUpdate
+    : dateFrom;
+
+  // Get historic transactions from backend API
   const history = await getTransactionsHistoryForAddresses(
     groupOfAddresses,
     updatedDateFrom
   );
-  if (history.length > 0) {
-    // FIXME: Add an endpoint for querying the best_block_num
-    // Update last block, done for one tx as all the best_block_num of a request are the same
-    const lastKnownBlockNumber = getLastBlockNumber();
-    if (!lastKnownBlockNumber || history[0].best_block_num > lastKnownBlockNumber) {
-      saveLastBlockNumber(history[0].best_block_num);
-    }
-
-    const transactions = previousTxs.concat(
-      _mapToAdaTxs(history, allAddresses)
+
+  // No more history left to fetch
+  if (history.length === 0) {
+    return previousTxs;
+  }
+
+  /* Update last block to best block in wallet history.
+   * Note: Done for one tx as the best_block_num is the same for all txs in request
+  */
+  const bestBlockNum = Number(history[0].best_block_num);
+  const lastKnownBlockNumber = getLastBlockNumber();
+  if (!lastKnownBlockNumber || bestBlockNum > lastKnownBlockNumber) {
```

Comment:
It’s 2am so I may say something that is not entirely well thought. Would this limit the ability to support rollbacks?

Reply:
The reason the best block is the same for all transactions is because the data is pulled from the DB by doing a `join` on the transaction table and the best_block_num table in the backend-service
```
FROM "txs"
  LEFT JOIN (SELECT * from "bestblock" LIMIT 1) f ON true
```
Since the best block table only contains a single row (the best block), it just gets copied in the join. It's really strange design to even join these tables but I think it's to reduce the amount of backend calls